lib/miq_expression.rb

```diff
@@ -155,13 +161,23 @@ def self._to_human(exp, options = {})
     clause
   end
 
-  def to_ruby(tz = nil)
-    # TODO: we cache the ruby expression regardless if the tz is different
-    if @ruby
+  # @param timezone [String]   (default: "UTC")
+  # @param prune_sql [boolean] (default: false) remove expressions that are sql friendly
+  #
+  # when prune_sql is true, then the sql friendly expression was used to filter the
+  #                         records already. no reason to do that again in ruby
+  def to_ruby(timezone = nil, prune_sql: false)
+    # TODO: base cache upon timezone and prune_sql values
+    # NOTE: not sure if the timezone or prune_sql will ever change between calls
+    if @ruby == true
+      nil
+    elsif @ruby
       @ruby.dup
     elsif valid?
-      @ruby ||= self.class._to_ruby(exp.deep_clone, context_type, tz || "UTC".freeze)
-      @ruby.dup
+      pexp = preprocess_exp!(exp.deep_clone)
+      pexp, _ = reduce_exp(pexp, MODE_RUBY) if prune_sql
+      @ruby = self.class._to_ruby(pexp, context_type, timezone || "UTC".freeze) || true
+      @ruby == true ? nil : @ruby.dup
```

Comment:
hmm, why was the `true` added?  Won't it be nil in that case anyway?

Reply:
The bigger concern is we're doubling down on caching options... timezone and now prune_sql. 😉

Reply:
@jrafanie great question.

if `_to_ruby` is a no-op, it returns `nil`. This means the code should evaluate this to always be `true`. (You can see `lenient_evaluate` checking `nil?` first)

Unfortunately, `nil` is the same as undefined.

So I wanted to set a value that means "there is no ruby for us to do, just assume the logical value for this is true."

I tried changing the constructor to default `@ruby` to `false` (meaning it was undefined). But we deserialize these from `Yaml`, and yaml does not use the constructor but rather just defaults the values to `nil`.

Reply:
> The bigger concern is we're doubling down on caching options... timezone and now prune_sql.

I did run into this issue when I was running benchmarks. I had to recreate the expression every time I changed the `prune_sql` value. I can't think of a reason why the code would pass this value.

I spent an hour looking for where we pass in timezones. The places that passed in a timezone derived the timezone from the report's time profile or the user. And those were only ever called from a single user.

I do remember working on code that populated reports that looked up all the timezones for the report group and generated report (values) for each timezone. I'm having trouble tracking this down though. But this is the only use case I can think of that would give the tz an issue.

Now as for the prune_sql. The only way this would be an issue is if we use a miqexpression in 2 places, one where we decide to use the expression to filter in sql and one where we don't. Feels very far fetched, but I guess it is possible.